File: src/main/java/architecture/goldenboughs_lib/core/LibEventHooks.java
package architecture.goldenboughs_lib.core;

import architecture.goldenboughs_lib.event.rationality.RationalityEvent;
import net.minecraft.world.entity.player.Player;
import net.neoforged.fml.ModLoader;

import java.util.Map;

public final class LibEventHooks {
	private LibEventHooks() {
	}

	public static Map.Entry<Boolean, Float> sourceRationalityPre(Player player, float oldValue, float newValue) {
		var event = new RationalityEvent.Modify.Pre(player, oldValue, newValue);
		ModLoader.postEvent(event);
		var canceled = event.isCanceled();
		return Map.entry(canceled,
			canceled ? event.getOldValue() : event.getNewValue());
	}

	public static void sourceRationalityPost(Player player, float oldValue, float newValue) {
		ModLoader.postEvent(new RationalityEvent.Modify.Post(player, oldValue, newValue));
	}
}

File: src/main/kotlin/architecture/goldenboughs_lib/core/LibEventHooks.kt
package architecture.goldenboughs_lib.core

import architecture.goldenboughs_lib.event.PlayerDropItemEvent
import architecture.goldenboughs_lib.event.PlayerHotbarChangeEvent
import architecture.goldenboughs_lib.module.rationality.event.RationalityEvent
import net.minecraft.world.InteractionHand
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.neoforged.neoforge.common.NeoForge

object LibEventHooks {
	@JvmStatic
	fun sourceRationalityPre(player: Player, oldValue: Float, newValue: Float): RationalityEvent.Modify.Pre {
		return NeoForge.EVENT_BUS.post(RationalityEvent.Modify.Pre(player, oldValue, newValue))
	}

	@JvmStatic
	fun sourceRationalityPost(player: Player, oldValue: Float, newValue: Float) {
		NeoForge.EVENT_BUS.post(RationalityEvent.Modify.Post(player, oldValue, newValue))
	}

	@JvmStatic
	fun playerHotbarChange(
		player: Player,
		fromSlot: Int,
		toSlot: Int,
		fromStack: ItemStack,
		toStack: ItemStack,
	): PlayerHotbarChangeEvent {
		return NeoForge.EVENT_BUS.post(PlayerHotbarChangeEvent(player, fromSlot, toSlot, fromStack, toStack))
	}

	@JvmStatic
	fun playerDropItem(
		player: Player,
		hand: InteractionHand,
	): PlayerDropItemEvent {
		return NeoForge.EVENT_BUS.post(PlayerDropItemEvent(player, hand))
	}
}
